Commit each created sample individually

ramonski opened the pull request:
## Description of the issue/feature this PR addresses

Under load (many concurrent operators creating samples), the sample
add form produced `ZODB.POSException.ConflictError` exceptions because
all rows of a submission committed in a single transaction and the
shared `INumberGenerator` storage (`OIBTree` on the portal annotation
`bika.lims.consecutive_numbers_storage`) became a hot contention
point.

This PR reworks the sample creation path in
`bika.lims.browser.analysisrequest.add2` so each sample is committed
in its own transaction with a per-sample retry on `ConflictError`,
and exposes the new behaviour as an opt-in registry flag.

## Current behavior before PR

- The whole submission is wrapped in one transaction.
- A conflict on any row aborts the entire batch.
- On a publisher-level retry the operator sees a generic error and
  has no easy way to identify which rows were affected.

## Desired behavior after PR is merged

- Each sample is created and committed individually.
- Transient `ConflictError`s are retried per sample with bounded
  exponential backoff.
- Partial failures are reported back to the user with enough
  information to recover the missing rows.
- The legacy single-transaction path stays available and remains
  the default.

## Implementation details

### Registry flag

A new boolean control panel field
`sample_add_form_commit_per_sample` (in the existing **Samples**
fieldset of the SENAITE registry, next to
`sample_add_form_allow_multi_paste` and
`trigger_events_on_sample_creation`) selects the strategy:

- **off (default)** — legacy `_create_samples_single_transaction`.
- **on** — `_create_samples_per_commit` with per-sample retry.

An upgrade step bumps the GenericSetup profile from `2727` to `2728`
and reuses the existing `import_registry` handler so existing sites
do not get a `KeyError` for the new record.

### Per-sample retry

`_create_one_with_retry` wraps `create_sample` with:

- `MAX_CREATE_ATTEMPTS = 8` (module-level constant).
- Exponential backoff capped at 2 s with jitter:
  `min(2.0, 0.05 * (2 ** attempt)) * uniform(0.5, 1.5)`.
- `transaction.begin()` / `transaction.commit()` per attempt;
  `transaction.abort()` on `ConflictError`.
- `source` propagation preserved for copy-to-new / partition flows.

After the first pass, samples that still failed are retried in a
short second pass (with a small randomised pause) so a brief burst
of contention doesn't surface to the user.

### Identifiable partial failures

Rows that ultimately fail are collected with their column index and
reported via a single portal warning message built by
`_report_failed_records`. The identifier shown per row prefers
`ClientSampleID`, falls back to `ClientReference`, and finally `-`
when the operator did not type any identifying field. Example:

```
1 sample(s) could not be created: #2 (CSID-A12), #5 (ref-9), #7 (-)
```

The message is a `zope.i18nmessageid.Message` with the row list in
its `mapping` so translations interpolate correctly.

### Publisher-retry idempotency

`/ar_add` is reachable through `zope.retry`, which can replay the
whole POST on a publisher-level `TransientError`. Because we now
commit per sample, a naive replay would create duplicates for every
row that was committed before the failure.

To prevent that, the per-commit path uses a worker-local idempotency
cache keyed by a SHA-256 fingerprint of
`REQUEST_METHOD + PATH_INFO + BODY`:

- `_inflight` — `OrderedDict` of `key -> (list_of_uids,
  monotonic_ts)`.
- `INFLIGHT_CACHE_SIZE = 32`, `INFLIGHT_CACHE_TTL = 600` s.
- Guarded by a `threading.Lock`.
- On hit, the previously created samples are resolved via
  `api.get_object_by_uid` and returned without re-creating anything.
- On miss, the list grows as each sample commits, so a mid-batch
  retry resumes where the previous attempt left off.

Request-annotation storage was considered and rejected: Zope
discards the `HTTPRequest` between retries and rebuilds it from the
WSGI environ, so request annotations don't survive a publisher
retry.

### Doctest

A new doctest `SampleCreationConflictRetry.rst` exercises the retry
path without requiring a real multi-connection ZODB conflict:

- Stubs `add2.transaction` and `add2.time` with fake module objects
  so the test-layer transaction and real `time.sleep` are not
  touched.
- Verifies `_record_identifier` precedence and fallback.
- Verifies that toggling the registry flag dispatches to the right
  internal path.
- Drives `_create_one_with_retry` with a flaky `create_sample` stub
  that raises `ConflictError` on the first call, then asserts that
  the helper retries and returns the sample on the second call.
- Drives the helper with an always-failing stub and asserts the
  exhaustion path returns `None` after `MAX_CREATE_ATTEMPTS`.
- Stubs `plone_utils.addPortalMessage` and asserts the
  `Message.mapping["rows"]` produced by `_report_failed_records`.

## Files touched

- `src/bika/lims/browser/analysisrequest/add2.py` — main logic, retry
  helpers, idempotency cache, dispatch.
- `src/senaite/core/registry/schema.py` — new
  `sample_add_form_commit_per_sample` field.
- `src/senaite/core/profiles/default/metadata.xml` — profile version
  bump (2727 → 2728).
- `src/senaite/core/upgrade/v02_07_000.zcml` — new upgrade step on
  top.
- `src/senaite/core/tests/doctests/SampleCreationConflictRetry.rst` —
  new doctest.
- `CHANGES.rst` — entry for #2741.

--
I confirm I have tested this PR thoroughly and coded it according to
[PEP8][1] and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

ramonski committed:
Commit each created sample individually

ramonski commented:
Questions and things to check more deeply:

- We use `IIndexQueueProcessor` adapters to catalog new objects using a [before commit hook](transaction.addBeforeCommitHook).
Are already catalogued objects also reverted when a transaction error occurs?
- How to deal with eventual skipped Samples, e.g. if the max. attempts are reached?
Currently, these are simply omitted and the user is not informed properly.
Probably it makes more sense to retry infinitely to create a sample in such a setup.

grulisco commented:
> * How to deal with eventual skipped Samples, e.g. if the max. attempts are reached?
>       Currently, these are simply omitted and the user is not informed properly.
>       Probably it makes more sense to retry infinitely to create a sample in such a setup.

A **hundred** failing retries for **one** sample seem to me indicating a serious problem, so I'd prefer to cancel the whole process at this stage.

ramonski commented:
> > * How to deal with eventual skipped Samples, e.g. if the max. attempts are reached?
> >   Currently, these are simply omitted and the user is not informed properly.
> >   Probably it makes more sense to retry infinitely to create a sample in such a setup.
> 
> A **hundred** failing retries for **one** sample seem to me indicating a serious problem, so I'd prefer to cancel the whole process at this stage.

It depends on how many people are creating samples or writing into the same catalogs or backrefs at the same time, we might reach such high numbers of conflicts.

In my test I created 2x 50 samples "simultaneously" and noted more than 10 conflicts for some samples.

As we create new samples and only catch `ConflictError`s, I actually think that we might remove the retries completely.

xispa commented:
From a technical perspective, just a reminder we have the [`retriable` decorator](https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/decorators/__init__.py#L44) :wink:

ramonski commented:
> From a technical perspective, just a reminder we have the [`retriable` decorator](https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/decorators/__init__.py#L44) 😉

Oh, great! I completely forgot about this one.
I'll extend this decorator with the logging and refactor the code accordingly.

ramonski committed:
Merge branch '2.x' into sample-create-single-transactions

ramonski committed:
removed conflict marker

ramonski committed:
Refactored creation flow

ramonski committed:
Two-pass retry with identifiable failures in sample creation

Records that hit a ConflictError on the first per-sample commit are
kept aside and retried in a second pass once the initial burst has
settled. Records that still fail are surfaced with their column index
and any operator-supplied identifier (ClientSampleID, ClientReference)
in the warning portal message, and the full record is dumped to the
log for support. Each ConflictError now also logs the contended
class_name and oid so counter, container and catalog contention can
be told apart in production.

ramonski committed:
Make per-sample commit configurable in the registry

Add sample_add_form_commit_per_sample to the existing Samples fieldset
of ISenaiteRegistry. When enabled, create_samples dispatches to the
per-sample commit path with retry-on-conflict and second-pass retry;
when disabled (the default), the legacy single-transaction loop runs,
keeping the existing all-or-nothing semantics for sites that depend on
them.

ramonski committed:
Add upgrade step for per-sample commit registry record + doctest

Existing instances upgrading to this version need the new
`sample_add_form_commit_per_sample` registry record to exist before
the Samples control panel is rendered, otherwise plone.registry
raises a KeyError. Bump the GenericSetup metadata version and add an
upgrade step that re-imports `plone.app.registry` from the senaite.core
profile to create the record.

Add a doctest at SampleCreationConflictRetry.rst that exercises the
identifier helpers, the toggle dispatch between the legacy single-
transaction path and the per-sample-commit path, the retry-on-conflict
behaviour using a stubbed `create_sample`, the failed-record portal
message, and the property that aborted ID-counter writes do not burn
sequence numbers.

ramonski committed:
Fix SampleCreationConflictRetry doctest

The previous version drove `transaction.begin/commit/abort` against
the test-layer connection, which detached the BTreeFolder used for the
`client` object on subsequent steps, queried the wrong API to extract
the failed-record portal message, and reset its own counter seed via
the same transaction handling.

Stub the `transaction` and `time` bindings on the add2 module so the
retry helper does not touch the layer's transaction, capture the
portal message directly via a stubbed `plone_utils`, and assert
against the i18n `Message.mapping["rows"]` value (the rows are not
substituted into the default string until rendering).

ramonski committed:
Reuse existing import_registry handler for the 2728 upgrade step

Drop the redundant import_registry_commit_per_sample handler and
point the 2727 -> 2728 step at the existing
v02_07_000.import_registry, which already re-imports the
plone.app.registry profile (with the AT typeinfo ordering workaround
documented in its body).

ramonski committed:
Drop CONCEPT_SAMPLE_DUPLICATE.md from this branch

Was accidentally included in the previous commit by `git add -A`;
unrelated internal planning doc and not part of this PR.

ramonski committed:
Changelog entry for #2741

ramonski committed:
Move MAX_CREATE_ATTEMPTS to module level

Project convention: module-level constants live at the top below
imports, not as class attributes. Promote MAX_CREATE_ATTEMPTS next to
the other add2 constants and update the retry helper plus the doctest
to reference it via the module.

ramonski committed:
Make per-sample commit idempotent on publisher retry

Per-sample commits during ar_add submission are durable before the
WSGI publisher commits the response transaction. If that final
commit hits a ConflictError (typically on the IStatusMessage
OIBucket from `addPortalMessage`, or on a catalog bucket touched
during reindexing) Zope's transaction_pubevents catches it and
re-publishes the request. Without state, the re-publish would call
create_samples again on the same form data and create duplicates
for every record we had already committed.

Add a worker-local in-flight cache keyed by a content fingerprint of
the submission body (method + path + raw body, SHA-256). Each
successful per-sample commit appends the new sample UID to a list
held in the cache entry; a concurrent retry on the same worker hits
the cache, looks up the durable samples by UID, and returns them
without re-creation.

The cache is an LRU OrderedDict capped at 32 entries with a 10
minute TTL. Eviction is lazy on access; the cap is defensive
(steady-state size is the number of concurrent submissions on this
worker, typically 1-5). State is process-local because Zope's retry
path stays on the same worker thread and rebuilds the request from
the same environ, so we never need cross-process visibility.

The legacy single-transaction path is untouched.

xispa reviewed (approved):
Really nice improvement, thanks!
